include/lbann/io/persist.hpp

```diff
@@ -145,6 +206,101 @@ bool read_double (int fd, const char *name, double *val);
 bool write_string(int fd, const char *name, const char *buf, size_t size);
 bool read_string(int fd, const char *name, char *buf, size_t size);
 
+class NonexistentArchiveFile : public std::runtime_error {
+public:
+  NonexistentArchiveFile(std::string const& filename) : std::runtime_error(std::string("Archive file not found: ") + filename) {}
+};
+
+template <typename C>
+void write_cereal_archive(C& obj, persist& p, persist_type pt, const std::string& suffix) {
+  std::ofstream os(p.get_filename(pt) + suffix);
+  if(!os.is_open()) {
+    throw NonexistentArchiveFile("Unable to open file for saving checkpoint " + p.get_filename(pt) + suffix);
+  }
+  cereal::XMLOutputArchive archive(os);
+  archive(obj);
+  return;
+}
+
+template <typename C>
+void write_cereal_archive(C& obj, persist& p, execution_mode mode, const std::string& suffix) {
+  const persist_type pt = execution_mode_to_persist_type(mode);
+  write_cereal_archive<C>(obj, p, pt, suffix);
+  return;
+}
+
+template <typename C>
+void read_cereal_archive(C& obj, persist& p, persist_type pt, const std::string& suffix) {
+  std::ifstream is(p.get_filename(pt) + suffix);
+  if(!is.is_open()) {
+    throw NonexistentArchiveFile("Unable to open file for loading checkpoint " + p.get_filename(pt) + suffix);
+  }
+  cereal::XMLInputArchive archive(is);
+  archive(obj);
+  return;
+}
+
+template <typename C>
+void read_cereal_archive(C& obj, persist& p, execution_mode mode, const std::string& suffix) {
+  const persist_type pt = execution_mode_to_persist_type(mode);
+  read_cereal_archive<C>(obj, p, pt, suffix);
+  return;
+}
+
+template <typename C>
+std::string create_cereal_archive_binary_string(C& obj) {
+  std::ostringstream ss;
+  {
+    cereal::BinaryOutputArchive archive(ss);
+    archive(obj);
+  } // archive goes out of scope, ensuring all contents are flushed
+  return ss.str();
+}
+
+template <typename C>
+void unpack_cereal_archive_binary_string(C& obj, const std::string& buf) {
+  std::istringstream ss(buf);
+  {
+    cereal::BinaryInputArchive archive(ss);
+    archive(obj);
+  } // archive goes out of scope, ensuring all contents are flushed
+  return;
+}
+
+template <typename C>
+void load_from_shared_cereal_archive(C& obj, persist& p, persist_type pt,
+                                     lbann_comm& comm,
+                                     const std::string& suffix) {
+  std::string buf;
+  if (comm.am_trainer_master()) {
+    read_cereal_archive<C>(obj, p, pt, suffix);
+    buf = create_cereal_archive_binary_string<C>(obj);
+  }else {
+    // If you are not the trainer master, still check to see if the file exists
+    std::ifstream is(p.get_filename(pt) + suffix);
+    if(!is.is_open()) {
+      throw NonexistentArchiveFile("Unable to open file for loading checkpoint " + p.get_filename(pt) + suffix);
```

Comment:
1. Should be a filename, not an error message.
2. Can fail for reasons beyond the file not existing.